File: problems/3.2/research/scripts/q32_transverse_common_fast.py
#!/usr/bin/env python3
"""Sage driver for the fast exact Apéry transverse common-pair scan.

Run under Sage, for example

    sage -python problems/3.2/research/scripts/q32_transverse_common_fast.py all \
        --pmax 100000 --workers 16

The C++/FLINT worker makes every support decision modulo the prime being
scanned.  This driver compiles the worker, balances primes across processes,
resumes from per-prime atomic checkpoints, verifies the known r<=10000 high
pairs, and computes p-adic multiplicities of the hits by independent Sage
power-series arithmetic modulo p^e.

No floating-point number is used to decide a common pair or a multiplicity.
Floating point occurs only in workload balancing, elapsed-time reporting, and
human-readable p/r output.
"""

from __future__ import annotations

import argparse
from fractions import Fraction
import heapq
import json
import math
import os
from pathlib import Path
import shutil
import subprocess
import sys
from typing import Iterable, Sequence

try:
    from sage.all import Integers, PowerSeriesRing, ZZ, prime_range
except Exception as exc:  # pragma: no cover - diagnostic outside Sage
    raise SystemExit(
        "This driver needs Sage.  Run it as `sage -python "
        "problems/3.2/research/scripts/q32_transverse_common_fast.py ...`.\n"
        f"Import error: {exc}"
    )


SCRIPT = Path(__file__).resolve()
SCRIPT_DIR = SCRIPT.parent
CPP_SOURCE = SCRIPT.with_suffix(".cpp")
DEFAULT_BINARY = SCRIPT_DIR / ".build" / "q32_transverse_common_fast"
DEFAULT_CHECKPOINTS = SCRIPT_DIR / "q32_transverse_fast_checkpoints"
ALGO_VERSION = "q32-transverse-common-fast-v1"
KNOWN_R10000_HIGH = {(17, 13), (2237, 492)}


def apery_mod(modulus: int, N: int) -> list[int]:
    """b_0,...,b_{N-1} modulo modulus, assuming every 1,...,N is a unit."""
    if N <= 0:
        return []
    b = [0] * N
    b[0] = 1 % modulus
    if N == 1:
        return b
    b[1] = 5 % modulus
    for n in range(1, N - 1):
        n2 = n * n % modulus
        n3 = n2 * n % modulus
        coeff = (34 * n3 + 51 * n2 + 27 * n + 5) % modulus
        numerator = (coeff * b[n] - n3 * b[n - 1]) % modulus
        inv = pow(n + 1, -1, modulus)
        b[n + 1] = numerator * pow(inv, 3, modulus) % modulus
    return b


def inv_sqrt_D_mod(modulus: int, N: int) -> list[int]:
    """Coefficients of (1-34t+t^2)^(-1/2) modulo modulus."""
    if N <= 0:
        return []
    q = [0] * N
    q[0] = 1 % modulus
    if N == 1:
        return q
    q[1] = 17 % modulus
    for n in range(1, N - 1):
        numerator = ((34 * n + 17) * q[n] - n * q[n - 1]) % modulus
        q[n + 1] = numerator * pow(n + 1, -1, modulus) % modulus
    return q


def xi_residue_mod_prime_power(p: int, r: int, exponent: int) -> tuple[int, int]:
    """Return (b_r, Xi_r) modulo p^exponent by an independent Sage path.

    Since r<p for every scanner hit, all recurrence denominators are units
    modulo p^e.  F(0)=1 makes the power-series inverse valid over Z/p^e Z.
    """
    if not (0 <= r < p):
        raise ValueError("valuation verifier requires r < p")
    modulus = int(p) ** int(exponent)
    N = r + 1
    b = apery_mod(modulus, N)
    q = inv_sqrt_D_mod(modulus, N)
    ring = Integers(modulus)
    series = PowerSeriesRing(ring, "t", default_prec=N)
    F = series([ring(x) for x in b]).add_bigoh(N)
    Q = series([ring(x) for x in q]).add_bigoh(N)
    Finv = (~F).add_bigoh(N)
    G = (Finv * Finv * Q).add_bigoh(N)
    xi = ring(-1)
    for m in range(1, N):
        xi -= ring(5) * G[m] * ring(b[m - 1])
    return int(ring(b[r])), int(xi)


def residue_valuation(residue: int, p: int) -> int:
    if residue == 0:
        raise ValueError("zero residue does not determine a finite valuation")
    v = 0
    while residue % p == 0:
        residue //= p
        v += 1
    return v


def exact_hit_valuation(p: int, r: int) -> dict[str, int]:
    """Determine v_p(b_r), v_p(Xi_r), and their minimum exactly.

    The exponent doubles until both residues are nonzero.  b_r is a positive
    integer and Xi_r=-1-5*sum_{m<=r} g_m b_{m-1}<0 over Z, so both valuations
    are finite; the loop therefore terminates.  Once x mod p^e is nonzero,
    with v_p(x)<e, its residue has exactly the true p-adic valuation.
    """
    exponent = 2
    vb = None
    vx = None
    while vb is None or vx is None:
        br, xir = xi_residue_mod_prime_power(p, r, exponent)
        if vb is None and br != 0:
            vb = residue_valuation(br, p)
        if vx is None and xir != 0:
            vx = residue_valuation(xir, p)
        exponent *= 2
    return {"vp_b": int(vb), "vp_xi": int(vx), "vp_gcd": int(min(vb, vx))}


def self_test() -> None:
    """Small exact characteristic-zero test of the formulas used by both paths."""
    b = [ZZ(1), ZZ(5)]
    for n in range(1, 2):
        numerator = (34 * n**3 + 51 * n**2 + 27 * n + 5) * b[n] - n**3 * b[n - 1]
        denominator = (n + 1) ** 3
        assert numerator % denominator == 0
        b.append(numerator // denominator)
    assert b == [1, 5, 73]

    q = [ZZ(1), ZZ(17)]
    for n in range(1, 2):
        numerator = (34 * n + 17) * q[n] - n * q[n - 1]
        assert numerator % (n + 1) == 0
        q.append(numerator // (n + 1))
    assert q == [1, 17, 433]

    S = PowerSeriesRing(ZZ, "t", default_prec=3)
    F = S(b).add_bigoh(3)
    Q = S(q).add_bigoh(3)
    G = ((~F) * (~F) * Q).add_bigoh(3)
    assert [ZZ(G[i]) for i in range(3)] == [1, 7, 192]
    xi = ZZ(-1)
    xis = [xi]
    for m in range(1, 3):
        xi -= 5 * ZZ(G[m]) * b[m - 1]
        xis.append(xi)
    assert xis == [-1, -36, -4836]


def compile_worker(binary: Path, *, force: bool = False) -> None:
    binary = binary.resolve()
    binary.parent.mkdir(parents=True, exist_ok=True)
    if binary.exists() and not force and binary.stat().st_mtime >= CPP_SOURCE.stat().st_mtime:
        return

    cxx = os.environ.get("CXX") or shutil.which("g++") or shutil.which("c++")
    if not cxx:
        raise SystemExit("No C++ compiler found (set CXX or install g++).")
    cmd = [cxx, "-O3", "-DNDEBUG", "-std=c++17", str(CPP_SOURCE), "-o", str(binary)]

    sage_local = os.environ.get("SAGE_LOCAL")
    if sage_local:
        include = Path(sage_local) / "include"
        lib = Path(sage_local) / "lib"
        cmd += [f"-I{include}", f"-L{lib}", f"-Wl,-rpath,{lib}"]
    else:
        try:
            flags = subprocess.check_output(
                ["pkg-config", "--cflags", "--libs", "flint"], text=True
            ).split()
            cmd += flags
        except Exception:
            pass
    cmd += ["-lflint", "-lgmp", "-lmpfr"]
    print("BUILD", " ".join(map(str, cmd)), flush=True)
    subprocess.run(cmd, check=True)


def read_checkpoint(path: Path) -> dict:
    data = json.loads(path.read_text())
    if data.get("version") != ALGO_VERSION:
        raise RuntimeError(f"checkpoint {path} has wrong version: {data.get('version')}")
    return data


def completed_primes(checkpoint_dir: Path) -> set[int]:
    done = set()
    if not checkpoint_dir.exists():
        return done
    for path in checkpoint_dir.glob("p_*.json"):
        try:
            data = read_checkpoint(path)
        except Exception:
            continue
        done.add(int(data["p"]))
    return done


def balanced_bins(primes: Sequence[int], workers: int) -> list[list[int]]:
    """Greedy LPT partition using p*log(p) as a conservative FLINT cost proxy."""
    workers = max(1, min(workers, len(primes) or 1))
    heap: list[tuple[float, int, list[int]]] = [(0.0, i, []) for i in range(workers)]
    heapq.heapify(heap)
    for p in sorted(primes, reverse=True):
        load, i, values = heapq.heappop(heap)
        values.append(p)
        load += p * max(1.0, math.log2(p))
        heapq.heappush(heap, (load, i, values))
    bins = [[] for _ in range(workers)]
    while heap:
        _load, i, values = heapq.heappop(heap)
        bins[i] = sorted(values)
    return [values for values in bins if values]


def run_scan(binary: Path, checkpoint_dir: Path, pmax: int, workers: int) -> None:
    if pmax < 2:
        return
    all_primes = [int(p) for p in prime_range(2, pmax + 1)]
    done = completed_primes(checkpoint_dir)
    pending = [p for p in all_primes if p not in done]
    print(
        f"SCAN p<= {pmax}: total={len(all_primes)} done={len(done & set(all_primes))} "
        f"pending={len(pending)} workers={workers}",
        flush=True,
    )
    if not pending:
        return

    jobs_dir = checkpoint_dir / ".jobs"
    jobs_dir.mkdir(parents=True, exist_ok=True)
    processes: list[tuple[subprocess.Popen, Path]] = []
    for index, values in enumerate(balanced_bins(pending, workers)):
        job = jobs_dir / f"job_{index:03d}.txt"
        job.write_text("".join(f"{p}\n" for p in values))
        cmd = [
            str(binary),
            "--prime-file",
            str(job),
            "--pmax",
            str(pmax),
            "--checkpoint-dir",
            str(checkpoint_dir),
        ]
        processes.append((subprocess.Popen(cmd), job))

    failures = []
    for proc, job in processes:
        code = proc.wait()
        if code:
            failures.append((job, code))
    if failures:
        raise RuntimeError(f"worker failures: {failures}")


def collect_summary(checkpoint_dir: Path, pmax: int, *, allow_partial: bool = False) -> dict:
    expected_primes = [int(p) for p in prime_range(2, pmax + 1)]
    records: dict[int, dict] = {}
    for path in checkpoint_dir.glob("p_*.json"):
        try:
            row = read_checkpoint(path)
        except Exception:
            continue
        p = int(row["p"])
        if p <= pmax:
            records[p] = row

    missing = [p for p in expected_primes if p not in records]
    if missing and not allow_partial:
        raise RuntimeError(
            f"scan incomplete: {len(missing)} missing primes; first={missing[:10]}"
        )

    pairs = sorted(
        (p, int(r))
        for p, row in records.items()
        for r in row.get("pairs", [])
    )
    detailed = []
    for p, r in pairs:
        valuation = exact_hit_valuation(p, r)
        detailed.append(
            {
                "p": p,
                "r": r,
                "ratio": str(Fraction(p, r)),
                "ratio_float": p / r,
                "barrier_violation": p > 5 * r,
                **valuation,
            }
        )

    max_pair = None
    if pairs:
        p, r = max(pairs, key=lambda pr: Fraction(pr[0], pr[1]))
        max_pair = {
            "p": p,
            "r": r,
            "ratio": str(Fraction(p, r)),
            "ratio_float": p / r,
        }

    baseline_observed = {(p, r) for p, r in pairs if r <= 10000}
    baseline_checked = pmax >= 100000 and not missing
    baseline_ok = baseline_checked and baseline_observed == KNOWN_R10000_HIGH
    if baseline_checked and not baseline_ok:
        raise RuntimeError(
            "r<=10000 high-pair regression: "
            f"observed={sorted(baseline_observed)} expected={sorted(KNOWN_R10000_HIGH)}"
        )

    summary = {
        "version": ALGO_VERSION,
        "pmax": pmax,
        "prime_count_expected": len(expected_primes),
        "prime_count_complete": len(records),
        "missing_primes": missing,
        "pair_count": len(pairs),
        "pairs": detailed,
        "max_ratio_pair": max_pair,
        "barrier_violations": [row for row in detailed if row["barrier_violation"]],
        "known_r10000_reference": sorted([list(x) for x in KNOWN_R10000_HIGH]),
        "known_r10000_reference_checked": baseline_checked,
        "known_r10000_reference_ok": baseline_ok,
        "total_b_zero_count": sum(int(row.get("b_zero_count", 0)) for row in records.values()),
        "total_worker_seconds": sum(float(row.get("seconds", 0.0)) for row in records.values()),
    }
    return summary


def print_summary(summary: dict) -> None:
    print(json.dumps(summary, indent=2, sort_keys=True))
    if summary["max_ratio_pair"]:
        row = summary["max_ratio_pair"]
        print(
            f"MAX p/r: p={row['p']} r={row['r']} ratio={row['ratio']} "
            f"(~{row['ratio_float']:.12f})"
        )
    print("COMMON p>r PAIRS:")
    for row in summary["pairs"]:
        print(
            f"  p={row['p']} r={row['r']} p/r={row['ratio']} "
            f"vp(b)={row['vp_b']} vp(Xi)={row['vp_xi']} "
            f"vp(gcd)={row['vp_gcd']}"
        )
    if summary["barrier_violations"]:
        print("P>5R COUNTEREXAMPLES:")
        for row in summary["barrier_violations"]:
            print(f"  p={row['p']} r={row['r']} ratio={row['ratio']}")
    else:
        print("P>5R COUNTEREXAMPLES: none in the completed scan range")


def parse_args(argv: Sequence[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "command",
        nargs="?",
        choices=("build", "scan", "summarize", "all", "self-test"),
        default="all",
    )
    parser.add_argument("--pmax", type=int, default=100000)
    parser.add_argument("--workers", type=int, default=max(1, (os.cpu_count() or 2) // 2))
    parser.add_argument("--binary", type=Path, default=DEFAULT_BINARY)
    parser.add_argument("--checkpoint-dir", type=Path, default=DEFAULT_CHECKPOINTS)
    parser.add_argument("--force-build", action="store_true")
    parser.add_argument("--allow-partial", action="store_true")
    parser.add_argument("--summary-json", type=Path)
    return parser.parse_args(argv)


def main(argv: Sequence[str] | None = None) -> None:
    args = parse_args(sys.argv[1:] if argv is None else argv)
    self_test()
    if args.command == "self-test":
        print("self-test: OK")
        return
    if args.command in ("build", "scan", "all"):
        compile_worker(args.binary, force=args.force_build)
        if args.command == "build":
            return
    if args.command in ("scan", "all"):
        run_scan(args.binary.resolve(), args.checkpoint_dir.resolve(), args.pmax, args.workers)
    if args.command in ("summarize", "all"):
        summary = collect_summary(
            args.checkpoint_dir.resolve(), args.pmax, allow_partial=args.allow_partial
        )
        print_summary(summary)
        output = args.summary_json or (args.checkpoint_dir / "summary.json")
        output.parent.mkdir(parents=True, exist_ok=True)
        output.write_text(json.dumps(summary, indent=2, sort_keys=True) + "\n")
        print(f"summary written to {output}")


if __name__ == "__main__":
    main()

File: problems/3.2/research/scripts/q32_transverse_common_fast.cpp
// Exact high-prime scanner for Apéry transverse common pairs.
//
// Build (inside a Sage environment, or with system FLINT):
//   g++ -O3 -DNDEBUG -std=c++17 q32_transverse_common_fast.cpp \
//       -lflint -lgmp -lmpfr -o q32_transverse_common_fast
//
// The scanner is deliberately prime-by-prime.  For one prime p it first
// computes b_0,...,b_{p-1} in O(p) word operations.  If no b_r vanishes,
// there cannot be a common pair and no power-series arithmetic is done.
// Otherwise it truncates at one past the last zero of b and lets FLINT compute
//
//   g = F^{-2} (1 - 34 t + t^2)^{-1/2}
//
// modulo p using one inverse_series and two mullow operations.  Decisions are
// exact modulo p; floating point is used only for timing output.

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#if __has_include(<flint/nmod_poly.h>)
#include <flint/nmod_poly.h>
#elif __has_include(<nmod_poly.h>)
#include <nmod_poly.h>
#else
#error "FLINT nmod_poly.h not found"
#endif

namespace fs = std::filesystem;
using u64 = std::uint64_t;
using u128 = __uint128_t;

static constexpr const char *ALGO_VERSION = "q32-transverse-common-fast-v1";

static inline u64 add_mod(u64 a, u64 b, u64 p) {
    u64 c = a + b;
    return c >= p ? c - p : c;
}

static inline u64 sub_mod(u64 a, u64 b, u64 p) {
    return a >= b ? a - b : a + p - b;
}

static inline u64 mul_mod(u64 a, u64 b, u64 p) {
    return static_cast<u64>((static_cast<u128>(a) * b) % p);
}

static inline u64 cube_mod(u64 a, u64 p) {
    return mul_mod(mul_mod(a, a, p), a, p);
}

static bool is_prime(u64 n) {
    if (n < 2) return false;
    if ((n & 1U) == 0) return n == 2;
    for (u64 d = 3; d * d <= n; d += 2)
        if (n % d == 0) return false;
    return true;
}

static std::vector<u64> primes_up_to(u64 lo, u64 hi) {
    if (hi < 2 || hi < lo) return {};
    std::vector<bool> sieve(static_cast<std::size_t>(hi + 1), true);
    sieve[0] = false;
    if (hi >= 1) sieve[1] = false;
    for (u64 q = 2; q * q <= hi; ++q) {
        if (!sieve[static_cast<std::size_t>(q)]) continue;
        for (u64 m = q * q; m <= hi; m += q)
            sieve[static_cast<std::size_t>(m)] = false;
    }
    std::vector<u64> out;
    for (u64 p = std::max<u64>(2, lo); p <= hi; ++p)
        if (sieve[static_cast<std::size_t>(p)]) out.push_back(p);
    return out;
}

static std::vector<u64> primes_from_file(const fs::path &path) {
    std::ifstream in(path);
    if (!in) throw std::runtime_error("cannot open prime file: " + path.string());
    std::vector<u64> out;
    u64 p;
    while (in >> p) {
        if (!is_prime(p))
            throw std::runtime_error("non-prime in prime file: " + std::to_string(p));
        out.push_back(p);
    }
    std::sort(out.begin(), out.end());
    out.erase(std::unique(out.begin(), out.end()), out.end());
    return out;
}

struct PrimeRecord {
    u64 p = 0;
    u64 r_limit = 0;
    u64 b_zero_count = 0;
    u64 truncation = 0;
    std::vector<u64> pairs;
    double seconds = 0.0;
};

static fs::path checkpoint_path(const fs::path &dir, u64 p) {
    std::ostringstream name;
    name << "p_";
    name.width(6);
    name.fill('0');
    name << p << ".json";
    return dir / name.str();
}

static void write_checkpoint_atomic(const fs::path &dir, const PrimeRecord &r) {
    fs::create_directories(dir);
    const fs::path final = checkpoint_path(dir, r.p);
    const fs::path temp = final.string() + ".tmp";
    {
        std::ofstream out(temp, std::ios::trunc);
        if (!out) throw std::runtime_error("cannot write checkpoint: " + temp.string());
        out << "{\"version\":\"" << ALGO_VERSION << "\","
            << "\"p\":" << r.p << ','
            << "\"r_limit\":" << r.r_limit << ','
            << "\"b_zero_count\":" << r.b_zero_count << ','
            << "\"truncation\":" << r.truncation << ','
            << "\"pairs\":[";
        for (std::size_t i = 0; i < r.pairs.size(); ++i) {
            if (i) out << ',';
            out << r.pairs[i];
        }
        out << "],\"seconds\":" << r.seconds << "}\n";
        out.flush();
        if (!out) throw std::runtime_error("failed while writing checkpoint");
    }
    if (fs::exists(final)) fs::remove(temp);
    else fs::rename(temp, final);
}

// b_0,...,b_{p-1}.  All denominators (n+1)^3 are units because n+1 < p.
static std::vector<u64> apery_mod_prime(u64 p, std::vector<u64> &inv) {
    const std::size_t N = static_cast<std::size_t>(p);
    inv.assign(N, 0);
    if (p > 1) inv[1] = 1;
    for (u64 n = 2; n < p; ++n) {
        // p is prime and n < p, so p % n != 0 unless n=1.
        const u64 term = mul_mod(p / n, inv[static_cast<std::size_t>(p % n)], p);
        inv[static_cast<std::size_t>(n)] = term == 0 ? 0 : p - term;
    }

    std::vector<u64> b(N, 0);
    b[0] = 1 % p;
    if (p == 1) return b;
    b[1] = 5 % p;
    for (u64 n = 1; n + 1 < p; ++n) {
        const u64 nn = n % p;
        const u64 n2 = mul_mod(nn, nn, p);
        const u64 n3 = mul_mod(n2, nn, p);
        u64 coeff = mul_mod(34 % p, n3, p);
        coeff = add_mod(coeff, mul_mod(51 % p, n2, p), p);
        coeff = add_mod(coeff, mul_mod(27 % p, nn, p), p);
        coeff = add_mod(coeff, 5 % p, p);
        const u64 num = sub_mod(
            mul_mod(coeff, b[static_cast<std::size_t>(n)], p),
            mul_mod(n3, b[static_cast<std::size_t>(n - 1)], p), p);
        const u64 iv = inv[static_cast<std::size_t>(n + 1)];
        const u64 iv3 = cube_mod(iv, p);
        b[static_cast<std::size_t>(n + 1)] = mul_mod(num, iv3, p);
    }
    return b;
}

// q_n for Q(t)=(1-34t+t^2)^(-1/2):
// (n+1) q_{n+1} = (34n+17) q_n - n q_{n-1}.
static std::vector<u64> inv_sqrt_D_coeffs(
    u64 p, std::size_t L, const std::vector<u64> &inv) {
    std::vector<u64> q(L, 0);
    if (L == 0) return q;
    q[0] = 1 % p;
    if (L == 1) return q;
    q[1] = 17 % p;
    for (u64 n = 1; n + 1 < L; ++n) {
        const u64 a = (34 * (n % p) + 17) % p;
        const u64 num = sub_mod(
            mul_mod(a, q[static_cast<std::size_t>(n)], p),
            mul_mod(n % p, q[static_cast<std::size_t>(n - 1)], p), p);
        q[static_cast<std::size_t>(n + 1)] =
            mul_mod(num, inv[static_cast<std::size_t>(n + 1)], p);
    }
    return q;
}

static PrimeRecord scan_prime(u64 p) {
    const auto start = std::chrono::steady_clock::now();
    PrimeRecord rec;
    rec.p = p;
    rec.r_limit = p - 1;

    if (p == 2) {
        // b_1 = 1 mod 2, so there is no r<p with b_r=0.
        rec.seconds = 0.0;
        return rec;
    }

    std::vector<u64> inv;
    std::vector<u64> b = apery_mod_prime(p, inv);
    std::vector<u64> zero_positions;
    for (u64 r = 1; r < p; ++r)
        if (b[static_cast<std::size_t>(r)] == 0) zero_positions.push_back(r);
    rec.b_zero_count = static_cast<u64>(zero_positions.size());

    if (zero_positions.empty()) {
        const auto stop = std::chrono::steady_clock::now();
        rec.seconds = std::chrono::duration<double>(stop - start).count();
        return rec;
    }

    const u64 max_zero = zero_positions.back();
    const std::size_t L = static_cast<std::size_t>(max_zero + 1);
    rec.truncation = static_cast<u64>(L);
    std::vector<u64> q = inv_sqrt_D_coeffs(p, L, inv);

    nmod_poly_t F, Q, Finv, Finv2, G;
    nmod_poly_init2(F, static_cast<ulong>(p), static_cast<slong>(L));
    nmod_poly_init2(Q, static_cast<ulong>(p), static_cast<slong>(L));
    nmod_poly_init2(Finv, static_cast<ulong>(p), static_cast<slong>(L));
    nmod_poly_init2(Finv2, static_cast<ulong>(p), static_cast<slong>(L));
    nmod_poly_init2(G, static_cast<ulong>(p), static_cast<slong>(L));

    for (std::size_t i = 0; i < L; ++i) {
        nmod_poly_set_coeff_ui(F, static_cast<slong>(i), static_cast<ulong>(b[i]));
        nmod_poly_set_coeff_ui(Q, static_cast<slong>(i), static_cast<ulong>(q[i]));
    }

    // F(0)=Q(0)=1, so all series inversions below are unconditionally valid.
    nmod_poly_inv_series(Finv, F, static_cast<slong>(L));
    nmod_poly_mullow(Finv2, Finv, Finv, static_cast<slong>(L));
    nmod_poly_mullow(G, Finv2, Q, static_cast<slong>(L));

    // Xi_0=-1 and Xi_r-Xi_{r-1}=-5 g_r b_{r-1}.
    u64 xi = p - 1;
    std::size_t zindex = 0;
    for (u64 m = 1; m <= max_zero; ++m) {
        const u64 gm = static_cast<u64>(
            nmod_poly_get_coeff_ui(G, static_cast<slong>(m)));
        const u64 delta = mul_mod(5 % p,
            mul_mod(gm, b[static_cast<std::size_t>(m - 1)], p), p);
        xi = sub_mod(xi, delta, p);
        if (zindex < zero_positions.size() && zero_positions[zindex] == m) {
            if (xi == 0) rec.pairs.push_back(m);
            ++zindex;
        }
    }

    nmod_poly_clear(F);
    nmod_poly_clear(Q);
    nmod_poly_clear(Finv);
    nmod_poly_clear(Finv2);
    nmod_poly_clear(G);

    const auto stop = std::chrono::steady_clock::now();
    rec.seconds = std::chrono::duration<double>(stop - start).count();
    return rec;
}

struct Options {
    u64 pmin = 2;
    u64 pmax = 100000;
    fs::path prime_file;
    fs::path checkpoint_dir = "q32_transverse_fast_checkpoints";
    bool force = false;
};

static Options parse_options(int argc, char **argv) {
    Options o;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        auto need = [&](const char *flag) -> std::string {
            if (i + 1 >= argc) throw std::runtime_error(std::string("missing value after ") + flag);
            return argv[++i];
        };
        if (a == "--pmin") o.pmin = std::stoull(need("--pmin"));
        else if (a == "--pmax") o.pmax = std::stoull(need("--pmax"));
        else if (a == "--prime-file") o.prime_file = need("--prime-file");
        else if (a == "--checkpoint-dir") o.checkpoint_dir = need("--checkpoint-dir");
        else if (a == "--force") o.force = true;
        else if (a == "--help" || a == "-h") {
            std::cout
                << "usage: q32_transverse_common_fast [--pmin N] [--pmax N]\n"
                << "       [--prime-file FILE] [--checkpoint-dir DIR] [--force]\n";
            std::exit(0);
        } else throw std::runtime_error("unknown argument: " + a);
    }
    if (o.pmax > 100000000ULL)
        throw std::runtime_error("refusing pmax>1e8: this scanner is optimized for word primes");
    return o;
}

int main(int argc, char **argv) {
    try {
        const Options opt = parse_options(argc, argv);
        fs::create_directories(opt.checkpoint_dir);
        std::vector<u64> primes = opt.prime_file.empty()
            ? primes_up_to(opt.pmin, opt.pmax)
            : primes_from_file(opt.prime_file);

        std::size_t done = 0;
        for (u64 p : primes) {
            if (p < opt.pmin || p > opt.pmax) continue;
            const fs::path cp = checkpoint_path(opt.checkpoint_dir, p);
            if (!opt.force && fs::exists(cp)) {
                ++done;
                continue;
            }
            PrimeRecord rec = scan_prime(p);
            write_checkpoint_atomic(opt.checkpoint_dir, rec);
            ++done;
            if (!rec.pairs.empty()) {
                for (u64 r : rec.pairs)
                    std::cout << "PAIR p=" << p << " r=" << r
                              << " ratio=" << static_cast<double>(p) / r << "\n";
                std::cout.flush();
            }
            if ((done % 100) == 0) {
                std::cerr << "processed " << done << "/" << primes.size()
                          << " primes; last p=" << p << "\n";
            }
        }
        return 0;
    } catch (const std::exception &e) {
        std::cerr << "fatal: " << e.what() << "\n";
        return 2;
    }
}
